File: src/rpp.py
#!/usr/bin/python
#
# 2019 Bernd Pfrommer
#
# This is test code written to implement
#
# "Robust Pose Estimation from a Planar Target"
# by Gerald Schweighofer and Alex Pinz

# I believe I found some errors in their paper, see documentation
# in rpp.cpp, the C++ source code.

import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
from tf.transformations import *
import math

"""
import numpy as np
import matplotlib.pyplot as plt
import rpp
%load_ext autoreload
%autoreload 2
plt.ion()  # switch interactive mode on
"""

def make_wph(w=1.0, h=2.0):
    """ make homogeneous world points for rectangular tag of
        width w and length l """
    wph = np.matrix([[-w,  h, 0, 1],
                     [ w,  h, 0, 1],
                     [ w, -h, 0, 1],
                     [-w, -h, 0, 1]])
    return wph

def plot_image_points(ip1h, ip2h=None, ip3h=None):
    """ plot multiple homogeneous image points as scatter plot"""
    plt.scatter(np.squeeze(np.array(ip1h[:,0])),
                np.squeeze(np.array(ip1h[:,1])),c='blue')
    if not (ip2h is None):
        plt.scatter(np.squeeze(np.array(ip2h[:,0])),
                    np.squeeze(np.array(ip2h[:,1])),c='red')
    if not (ip3h is None):
        plt.scatter(np.squeeze(np.array(ip3h[:,0])),
                    np.squeeze(np.array(ip3h[:,1])),c='green')
    plt.gca().axis('equal')
    plt.show()

def make_tf(beta=0, alpha=0, gamma=0, d=3.0):
    """ make transform T = R_z(gamma) * R_y(beta) * R_x(alpha) """
    T = np.matrix(rotation_matrix(gamma, (0,0,1))) * \
        np.matrix(rotation_matrix(beta,  (0,1,0))) * \
        np.matrix(rotation_matrix(alpha, (1,0,0)))
    T[0:3,3] = np.matrix([[0, 0, d]]).T
    return T

def project_points(wph_cam):
    """ project homogeneous world points in camera coord to image points """
    return (wph_cam/wph_cam[:,2])[:,0:3]

def V_from_v(v):
    """ make V matrices from rows of v, see paper """
    V = np.zeros((v.shape[0], v.shape[1], v.shape[1]))
    for i in range(0,v.shape[0]):
        V[i,:,:] = v[i,:].T*v[i,:]/(v[i,:]*v[i,:].T)
    return V

def make_homogeneous(x):
    """ make homogeneous coordinates by appending 1 """
    if len(x.shape) == 1:
        x = np.expand_dims(x, axis=0)
    xh = np.hstack((x, np.ones((x.shape[0], 1))))
    return xh

def rotate_to_z(translat):
    """ compute rotation to rotate optical axis to point to tag center"""
    z = np.array([0, 0, 1])
    t = np.squeeze(np.asarray(translat))
    t_norm = np.linalg.norm(t)
    txz = np.cross(t, np.asarray(z))
    sin_a_t_norm = np.linalg.norm(txz)
    if abs(sin_a_t_norm) < 1e-8:
        return np.matrix(np.eye(4))
    sin_a = sin_a_t_norm / t_norm
    n = txz / sin_a_t_norm
    R = np.matrix(rotation_matrix(math.asin(sin_a), n))
    return R

def decompose_R1Tilde(R1T):
    """ decompose \tilde{R1}, see paper """
    # compute R1T = Rz * Ry * Rz0
    e  = euler_from_matrix(R1T, 'rzyz')
    Rz  = np.matrix(rotation_matrix(e[0], (0,0,1)))
    Ry  = np.matrix(rotation_matrix(e[1], (0,1,0)))
    Rz0 = np.matrix(rotation_matrix(e[2], (0,0,1)))
    # RzTilde.inverse() = Rz0
    return Rz, Ry, Rz0, e[0], e[1]

def eos(wph, iph, T):
    """ computes error function E_os for given world, img points,
        and transform"""
    V = V_from_v(iph)
    wp_trans = (T * wph.T)[0:3,:]
    I = np.matrix(np.eye(3))
    E = 0
    for i in range(0,wph.shape[0]):
        err = (I - V[i,:,:])*wp_trans[:,i]
        E = E + err.T*err
    return E[0,0]

def compute_G(Vt):
    """ computes constant G from the paper (with fixes, not original!)"""
    n   = Vt.shape[0]
    ImVs = np.matrix(np.zeros((3,3)))
    for i in range(0, n):
        ImV = np.matrix(np.eye(3)) - Vt[i,:,:]
        ImVs = ImVs + ImV.T * ImV
    return np.linalg.inv(ImVs)

def make_K(pt):
    """ The matrix K matrix is defined as
          K*[1, beta_t, beta_t^2]^T = R_y(beta_t)*p"""
    p = np.squeeze(np.array(pt))
    return np.matrix([[p[0], 2*p[2],  -p[0]],
                      [p[1], 0,        p[1]],
                      [p[2], -2*p[0], -p[2]]])

def compute_FTF(Vt, Rz, pt):
    """ computes F^T * F where
        F^T * F = sum_i F_i^T * F
    and
        F_i = (I-V_i)*Rz(gamma)*K_i -(I-Vi)*G*[sum_j(I-V_j)^2 Rz(gamma) K_i]"""
    n   = pt.shape[0]
    G   = compute_G(Vt)

    C   = np.zeros((n, 3, 3))
    Csum = np.matrix(np.zeros((3, 3)))
    for i in range(0, Vt.shape[0]):
        ImVi = np.matrix(np.eye(3)) - Vt[i,:,:]
        C[i,:,:] = ImVi * np.matrix(Rz) * make_K(pt[i,:])
        Csum = Csum + ImVi.T * C[i,:,:]

    FTF = np.matrix(np.zeros((3,3)))
    for i in range(0, Vt.shape[0]):
        ImVi = np.matrix(np.eye(3)) - Vt[i,:,:]
        Fi = np.matrix(C[i,:,:]) - ImVi * G * Csum
        FTF = FTF + Fi.T * Fi
    return FTF

def compute_polynomial(FTF):
    """
    E_os(beta_t) = (1+beta_t^2)^{-2} * mu^T * (F^T * F) * mu
    where mu = [1, beta_t, beta_t^2]^T
    
    now express E_os(beta_t) and derivates as polynomials in beta_t
    
    E_os   = (1+beta^2)^{-2} * (sum_{i=0^n} f[n-i] beta^i)
    E_os'  = (1+beta^2)^{-3} * (sum_{i=0^n} g[n-i] beta^i)
    E_os'' = (1+beta^2)^{-4} * (sum_{i=0^n} h[n-i] beta^i)

    NOTE: ordering is opposite to the c++ code!!!
          i.e. f[0] is x^4 order in python!!!
    """
    # zeroth order deriv poly
    f = np.asarray([FTF[2, 2],
                    FTF[1, 2] + FTF[2, 1],
                    FTF[0, 2] + FTF[1, 1] + FTF[2, 0],
                    FTF[0, 1] + FTF[1, 0],
                    FTF[0, 0]])
    # first deriv polynomial
    g = np.asarray([-f[1], 4*f[0]-2*f[2], 3*(f[1]-f[3]), 2*f[2]-4*f[4], f[3]])

    # second deriv polynomial
    h = np.asarray([2*f[1],            # 5
                    6 *f[2] - 12*f[0], # 4
                    12*f[3] - 16*f[1], # 3
                    20*f[4] - 16*f[2] + 12*f[0], # 2
                    -12*f[3] + 6*f[1], # 1
                    -4*f[4] +  2*f[2]])
    return f, g, h

def compute_optimal_t(Vt, Rz, beta, pt):
    """
    Computes optimal translation \tilde{t} for given beta.
    only used for testing
    """
    t  = np.matrix(np.zeros((3,1)))
    Ry = np.matrix(rotation_matrix(beta, (0,1,0)))
    G = np.matrix(np.zeros((3,3)))
    for j in range(0, Vt.shape[0]):
        ImV = np.matrix(np.eye(3)) - Vt[j]
        t = t + ImV * ImV * Rz[0:3,0:3] * Ry[0:3,0:3] * pt[j,0:3].T
        G = G + ImV * ImV
    return -np.linalg.inv(G) * t
    
def find_polynomials(wp, ip, T1):
    """
    Computes the polynomials f,g,h with 0th, 1st, and 2nd deriv of E_os:

    E_os   = (1+beta^2)^{-2} * (sum_{i=0^n} f[i] beta^i)
    E_os'  = (1+beta^2)^{-3} * (sum_{i=0^n} g[i] beta^i)
    E_os'' = (1+beta^2)^{-4} * (sum_{i=0^n} h[i] beta^i)
    """

    n   = wp.shape[0]   # number of points
    wph = make_homogeneous(wp)
    iph = make_homogeneous(ip)
    # compute rotation matrix Rt for use in equation (5)
    Rt  = rotate_to_z(np.squeeze(T1[0:3,3]))
    # R1T = R1Tilde, from equation 5
    R1T = Rt * T1
    Rz, Ry, Rz0, gamma_1, beta_1 = decompose_R1Tilde(R1T)
    vt  = (Rt[0:3,0:3] * iph.T).T  # v_tilde, equation (5)
    pt  = (Rz0 * wph.T).T # p_tilde, between equation 7/8
    Vt  = V_from_v(vt)   # V_tilde_i
    FTF = compute_FTF(Vt, Rz[0:3, 0:3], pt)
    f, g, h  = compute_polynomial(FTF)
    
    return f, g, h

def eos_from_poly(beta, f):
    """ compute E_os(beta)  from polynomial expression"""
    x = np.tan(beta/2.0)
    y = np.polyval(f, x)
    y = y / ((1.0 + x*x)*(1.0 + x*x))
    return y

def eos_deriv(beta, g):
    """ compute d E_os(beta)/d beta from polynomial expression"""
    x = np.tan(beta/2.0)
    y = g[4] + x * g[3] + x*x * g[2] + x*x*x*g[1] + x*x*x*x*g[0]
    y = y / ((1.0 + x*x)*(1.0 + x*x)*(1.0 + x*x))
    return y

def eos_from_full(x, pt, vt, Rz, Tt):
    """
    computes E_os(beta), not using polynomial, but also
    *without* optimizing t
    """
    y = np.zeros(x.shape[0])
    for i in range(0, x.shape[0]):
        Ry = np.matrix(rotation_matrix(x[i], (0,1,0)))
        R  = Rz * Ry
        R[0:3,3] = Tt[0:3,3]
        y[i] = eos(pt, vt, R)
    return y

def find_minima(f, g, h):
    """
    finds roots of g and then the real minima via h.
    returns E_os for all minima found
    """
    roots = np.roots(g)
    mins = []; E_os = [];
    for r in roots:
        if np.imag(r) < 1e-8 and np.polyval(h, r) > 0: # real, minimum
            beta_t = np.real(r)
            E = eos_from_poly(2.0*math.atan(beta_t), f)
            E_os.append(E)
            mins.append(2.0*math.atan(beta_t))
    return mins, E_os

def main(wp, ip, T1):
    """
    main driver routine that computes everything
    """
    wph = make_homogeneous(wp)
    iph = make_homogeneous(ip)
    # compute rotation matrix Rt for use in equation (5)
    Rt  = rotate_to_z(np.squeeze(T1[0:3,3]))
    # R1T = R1Tilde, from equation 5
    R1T = Rt * T1
    Rz, Ry, Rz0, gamma_1, beta_1 = decompose_R1Tilde(R1T)
    print 'beta_1: ', beta_1
    vt  = (Rt[0:3,0:3] * iph.T).T  # v_tilde, equation (5)
    pt  = (Rz0 * wph.T).T # p_tilde, between equation 7/8
    # Tt is the combination of Rz*Ry and t_tilde
    Tt  = Rz * Ry
    Tt[0:3,3] = np.squeeze(R1T[0:3,3]).T
    # all 3 of those should be identical
    print 'E_os original:                  ', eos(wph, iph, T1)
#    print 'E_os rotated once:  ', eos(wph, vt, R1T)
#    print 'E_os rotated twice: ', eos(pt, vt, Tt)

    f, g, h = find_polynomials(wp, ip, T1)
    betas, E_os = find_minima(f, g, h)
    print 'E_os from polynomial(beta_1):   ', eos_from_poly(beta_1, f)
    print 'E_os from polynomial(beta_min): ', eos_from_poly(betas[1], f)
#    print 'roots conv: ', 2*np.arctan(np.roots(g))
#    print 'second deriv: ', np.polyval(h, np.roots(g))
    print '--------- minima found -----------'
    for i in range(0, len(betas)):
        print 'E_os(', betas[i], ') = ', E_os[i]
    print 'ratio min/max: ', min(E_os)/max(E_os)
        
    beta = np.linspace(-math.pi, math.pi, 50)
    plt.plot(beta, eos_from_poly(beta, f), 'b')
#    plt.plot(beta, eos_deriv(beta, g), 'k')
#    plt.plot(beta, eos_from_full(beta, pt, vt, Rz, Tt),'r')
#    plt.show()
    return E_os, betas

File: src/rpp.cpp
/* -*-c++-*--------------------------------------------------------------------
 * 2019 Bernd Pfrommer bernd.pfrommer@gmail.com
 */

#include "tagslam/logging.h"
#include "tagslam/rpp.h"
#include "tagslam/quartic.h"
#include <boost/range/irange.hpp>
#include <iostream>
#include <math.h>

//
// Implementation of tests for checking if tag can be 'flipped',
// and a second valid pose can be obtained. This gives a measure of
// how well the tag-to-camera transform is established.
//
// See paper: "Robust Pose Estimation from a Planar Target"
// by Gerald Schweighofer and Alex Pinz
//
// Notes:
//
// 1) The paper is a bit nebulous about how to compute \tilde{R}_z^-1.
//    What it means concretely: decompose
//
//    \tilde{R}_1 = Rz(gamma) * Ry(beta) * Rz(gamma'),
//
//    and set \tilde{R}_z^{-1} = Rz(gamma'). Then 
//
//    \tilde{R}_1 * \tilde{R}_z = Rz(gamma) * Ry(beta),
//
//    i.e. only z and y rotations.
//    
// 2) top of page 6, beta_t = tan(1/(2*beta)) is wrong, it should
//    read beta_t = tan(beta/2)
//
// 3) I cannot reproduce their expression for the gradient. I think it's
//    wrong, but not sure. I derived my own, which works. This affects
//    equations (12) and (13) in the paper:
//
//    (12) becomes:
//
//    t_opt = -G * sum_i (I-\tilde{V}_i)^2 R_z(gamma) R_y(beta) \tilde{p}_i
//
//    with G = (sum_i (I-\tilde{V}_i)^2)^{-1}
//
//    (13) is modified accordingly:
//
//    E_os = sum_i |... - G sum_j(I -\tilde{V}_j)^2 R_z  R_y \tilde{p}_i|^2

// #define DEBUG

namespace tagslam {
  namespace rpp {
    using boost::irange;

    static double eval_poly(double x, const double *a, int n) {
      double p = 1.0;
      double sum = 0;
      for (const auto i: irange(0, n)) {
        sum += p * a[i];
        p = p * x;
      }
      return (sum);
    }

    //
    // computes rotation R_t from the paper: it rotates the optical axis
    // to face to the origin (center) of the tag.
    //
    static Transform rotate_to_z(
      const Eigen::Vector3d &translat, double *ang) {
      const double t_norm = translat.norm();
      const Eigen::Vector3d txz = translat.cross(Eigen::Vector3d::UnitZ());
      const double sin_a_t_norm = txz.norm();
      if (std::abs(sin_a_t_norm) < 1e-8) {
        return Eigen::Isometry3d::Identity();
      }
      const double sin_a = sin_a_t_norm / t_norm;
      const Eigen::Vector3d n = txz / sin_a_t_norm;
      //const Transform tf = Eigen::AngleAxisd(std::asin(sin_a), n);
      *ang = std::asin(sin_a);
      const Transform tf = (Transform) Eigen::AngleAxisd(*ang, n);
      return (tf);
    }

    typedef std::vector<Eigen::Matrix3d,
                        Eigen::aligned_allocator<Eigen::Matrix3d> > M33dVec;

    //
    // make normalized matrices \tilde{V}_i from \tilde{v}_i
    //
    static M33dVec V_from_v(const ImgPointsH &v) {
      M33dVec V(v.rows());
      for (int i = 0; i < v.rows(); i++) {
        const double vnsq = v.row(i).squaredNorm();
        if (vnsq > 1e-12) {
          V[i] = v.row(i).transpose()*v.row(i) / vnsq;
        } else {
          V[i] = Eigen::Matrix3d::Zero();
        }
      }
      return (V);
    }

    //
    // computes G from the paper (but using my formula)
    //
    static Eigen::Matrix3d compute_G(const M33dVec &V_tilde) {
      Eigen::Matrix3d ImV_sum = Eigen::Matrix3d::Zero();
      for (const auto &V_i: V_tilde) {
        const auto ImV = Eigen::Matrix3d::Identity() - V_i;
        ImV_sum = ImV_sum + ImV.transpose() * ImV;
      }
      return (ImV_sum.inverse());
    }

    //
    // little helper matrix K that extracts beta_t to various powers
    // from R_y(beta_t)*p
    //
    static Eigen::Matrix3d make_K(const Eigen::Vector3d &p) {
      Eigen::Matrix3d K;
      K <<
        p(0),  2*p(2), -p(0),
        p(1),     0.0,  p(1),
        p(2), -2*p(0), -p(2);
      return (K);
    }

    //
    // compute helper matrix FTF = F^T * F.
    //
    // The error E can be expressed then as:
    //
    // E_os(beta_t) = (1+beta_t^2)^{-2} * mu^T * (F^T * F) * mu
    //
    // where mu = [1, beta_t, beta_t^2]^T
    //

    static Eigen::Matrix3d compute_FTF(const M33dVec &V_tilde,
                                       const Transform &R_z,
                                       const ObjPoints &p_tilde) {
      const int n = p_tilde.rows(); // number of points
      const auto G = compute_G(V_tilde);
      Eigen::Matrix3d C_sum = Eigen::Matrix3d::Zero();
      M33dVec C(n);
      for (const auto i: irange(0ul, V_tilde.size())) {
        const auto ImV_i = Eigen::Matrix3d::Identity() - V_tilde[i];
        C[i] = ImV_i * R_z * make_K(p_tilde.row(i));
        C_sum = C_sum + ImV_i.transpose() * C[i];
      }
      Eigen::Matrix3d FTF = Eigen::Matrix3d::Zero();
      for (const auto i: irange(0ul, V_tilde.size())) {
        const auto ImV_i = Eigen::Matrix3d::Identity() - V_tilde[i];
        const auto F_i   = C[i] - ImV_i * G * C_sum;
        FTF = FTF + F_i.transpose() * F_i;
      }
      return (FTF);
    }

    //
    // Starting from
    //
    // E_os(beta_t) = (1+beta_t^2)^{-2} * mu^T * (F^T * F) * mu
    // where mu = [1, beta_t, beta_t^2]^T
    //
    // now express E_os(beta_t) and derivates as polynomials in beta_t
    //
    // E_os   = (1+beta^2)^{-2} * (sum_{i=0^n} f[i] beta^i)
    // E_os'  = (1+beta^2)^{-3} * (sum_{i=0^n} g[i] beta^i)
    // E_os'' = (1+beta^2)^{-4} * (sum_{i=0^n} h[i] beta^i)
    //
    //
    static void compute_polynomial(const Eigen::Matrix3d &FTF,
                                   double *f, double *g, double *h) {
      // polynomial coefficients f[0] == zeroth order etc
      //
      // E_os = (f[0] + f[1] * beta_t + ... f[4] * beta_t^4) / (1+beta_t^2)^2
      //
      f[0] = FTF(0, 0);
      f[1] = FTF(0, 1) + FTF(1, 0);
      f[2] = FTF(0, 2) + FTF(1, 1) + FTF(2, 0);
      f[3] = FTF(1, 2) + FTF(2, 1);
      f[4] = FTF(2, 2);

      // first derivate:
      g[4] = -f[3];
      g[3] = 4 * f[4] - 2 * f[2];
      g[2] = 3 * f[3] - 3 * f[1];
      g[1] = 2 * f[2] - 4 * f[0];
      g[0] = f[1];

      // second derivative:
      h[0] =  -4*f[0] +  2*f[2];
      h[1] = -12*f[1] +  6*f[3];
      h[2] =  20*f[0] - 16*f[2] + 12 * f[4];
      h[3] =  12*f[1] - 16*f[3];
      h[4] =   6*f[2] - 12*f[4];
      h[5] =   2*f[3];
    }

    //
    // Finds locations beta_t of real minima and value E there
    // Returns number of minima found.
    //

    static int find_minima(
      const double *f, // poly coeff for E_os
      const double *g, // poly coeff first deriv
      const double *h, // poly coeff second deriv
      double *beta_min, double *beta_max, double *E_min, double *E_max) { 
      std::complex<double> root[4];
      // find roots of first derivative
      const int nroots = quartic::solve_quartic(g[4],g[3],g[2],g[1],g[0],root);
      
      int n_min(0);
      *E_min = 1e90;
      *E_max = -1e90;
      // check all real roots, and evaluate second deriv there
      for (const auto i:irange(0, nroots)) {
        if (std::imag(root[i]) < 1e-8) {
          const double beta_t = std::real(root[i]);
          if (eval_poly(beta_t, h, 6) > 0) {
            const double opbs = (1.0 + beta_t * beta_t);
            const double E_os = eval_poly(beta_t, f, 5)/(opbs * opbs);
            if (E_os < *E_min) {
              *E_min = E_os;
              *beta_min = 2.0 * std::atan(beta_t);
            }
            if (E_os > *E_max) {
              *E_max = E_os;
              *beta_max = 2.0 * std::atan(beta_t);
            }
            n_min++;
          }
        }
      }
      return (n_min);
    }

    //
    // computes ratio of error for lowest/(second lowest) minimum error
    // orientation. The lower the ratio, the better is the pose
    // established, the more robust it is to flipping.
    //
    double check_quality(const ImgPoints &ip, const ObjPoints &op,
                         const Transform &T, double *beta_orig,
                         double *beta_min, double *beta_max) {
      const ImgPointsH iph = ip.rowwise().homogeneous();
      // compute R_t, the matrix that rotates z to the optical axis
      double ang;
      const Transform R_t = rotate_to_z(T.translation(), &ang);
      const auto R1_tilde = R_t * T;
      // decompose R1_tilde = Rz * Ry * Rz0
      const auto   euler_angles = R1_tilde.rotation().eulerAngles(2, 1, 2);
      const double gamma  = euler_angles[0]; // z  rotation
      const double beta   = euler_angles[1]; // y  rotation
      *beta_orig = beta;
      const double alpha  = euler_angles[2]; // z0 rotation
      const Transform R_z = (Transform)
        Eigen::AngleAxisd(gamma, Eigen::Vector3d::UnitZ());
      const Transform R_z0 = (Transform)
        Eigen::AngleAxisd(alpha, Eigen::Vector3d::UnitZ());
      // compute v_tilde from equation (5)
      const ImgPointsH v_tilde = (R_t.rotation()*iph.transpose()).transpose();
      // compute p_tilde
      const ObjPoints p_tilde  = (R_z0.rotation()*op.transpose()).transpose();
      M33dVec V_tilde = V_from_v(v_tilde);
      const Eigen::Matrix3d FTF = compute_FTF(V_tilde, R_z, p_tilde);
      double f[5], g[5], h[6];
      compute_polynomial(FTF, f, g, h);
      double E_min, E_max;
      int n_min = find_minima(f, g, h, beta_min, beta_max, &E_min, &E_max);
      switch (n_min) {
      case 2:
        return (E_min / E_max); // two minima, the usual case
        break;
      case 1:
        return (0.0); // single minimum, assume all is good
        break;
      default:
        ROS_WARN_STREAM("found bad num minima: " << n_min);
        *beta_min = beta;
        *beta_max = beta;
        return (0.0); // close eyes and hope for the best....
        break;
      }
      return (0.0); // should never reach this
    }
  } // end of namespace rpp
}
